### PR TITLE
Update CI policy: adding libraries should be justified + minor cleanups

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by ejgallego** [comment]
I wonder what is the rationale for the changes, in particular:

> For libraries: your project should exercise some part of Coq which is not well covered by the existing CI and test suite.

This indeed would make sense, but how do you plan to assess current and future coverage? For example, what's an example of a development that would be rejected under this new policy?

> ci/README-users: plugins don't have to compile with Coq's ocaml flags

This was added following the logic that plugins that are going to be maintained by Coq's team should follow the same coding and quality rules that Coq upstream. I still think (and experience in other codebases supports) that it is a good idea to try to enforce this.

**Comment by SkySkimmer** [comment]
Basically the reason we are not merging https://github.com/coq/coq/pull/19511 is that this is already the policy, I just try to make it more explicit.

> I still think (and experience in other codebases supports) that it is a good idea to try to enforce this.

It doesn't matter if it would be a good idea, nobody is doing it so let's stop pretending otherwise.

**Comment by andres-erbsen** [comment]
Would adding stronger language (and practice) about maintenance of CI-tracked libraries serve the need that this policy change is intended for? To me, predicting whether a library would find bugs seems extraordinarily challenging, as it is often hard to even agree whether a well-understood behavior change is a bug or not. For example, I doubt lambda-tamer would have been judged to meet the proposed policy, but last I checked it had bitrotted to multiple new anomalies. Applying the proposed criterion to new CI inclusions without revisiting already-included libraries seems also a little bit unfair, as I think those *were* accepted under the premise that CI inclusion is not contingent on the downstream's value as a test.

I do share the concern about maintenance burden of the CI, and I have previously [advocated for](https://github.com/coq/ceps/pull/86/files#diff-0750a9b2fc0dd02d0e9d66c491f3489d25e1c7f96539f4d23ec3c87a37ccd77bR129-R178) being more proactive about setting boundaries on it in a different manner. Updating libraries or plugins included in the CI should be the responsibility of their respective teams, with clear timeline expectations and allow-failure enforcement. Still, Coq developers would need to understand and document what kind of changes are expected to be needed (likely by porting their own downstreams). This is not a new concept (e.g. [here (2020)](https://github.com/coq/coq/commit/37070565e8b7bead15f5abda2c409f1ad2348fa1),) but I think we have been too polite to actually request downstreams to port their code over Coq changes. And when I ask myself why, the answer that comes to mind is that requesting downstream maintainers to work into porting would lead to minor embarrassment in case the breakage was in fact a new Coq bug, and that possibility does not seem all that unlikely for specific PR. Perhaps this is what you mean by "stop pretending otherwise", but I am still asking: are we sure we can't resolve this concern more directly?

I would also support more flexible guidance that encouraged avoiding duplication, and perhaps specifically running smaller test cases instead of larger ones on Coq CI as long as both exercise the same functionality, but an open-ended global uniqueness requirement goes too far IMO.

File: dev/ci/README-users.md
Comment (proux01) [inline dev/ci/README-users.md:20]: Maybe add a note that plugins are nowadays often advanteagously replaced by modern metalanguages such as Ltac2 or Elpi?

Comment (proux01) [reply dev/ci/README-users.md:20]: I mean, we don't want to encourage the proliferation of plugins either.

**Comment by proux01** [comment]
FWIW, I once tried to summarize the outcome of the discussion on CI contract launched by @palmskog during last working group in December in Nantes: https://github.com/proux01/coq/commit/ad58c2b3161efee206dc17f51a612d3a264a0237
Anyway, I support first merging this, considering it's a welcome clarification of the current state of affair.

**Comment by andres-erbsen** [comment]
My reading of the commit you linked is very different than the proposal here, "you can consider submitting your
project for addition to Coq's CI" and so on. Maybe something else was said in in-person discussions, but then it can surely be said again here so those who weren't present can be in on what changed and why.

**Comment by SkySkimmer** [comment]
Coq call: soften the message to mention capacity issues and encouraging to talk on zulip before submitting a PR

File: dev/ci/README-users.md
Comment (palmskog) [inline dev/ci/README-users.md:20]:
```suggestion
  There is no such requirement for plugins: developers may reduce the
  cost of working with Coq's unstable OCaml API by adding
  their plugin to Coq's CI and getting it fixed when the API changes (as long
  as fixing the plugin doesn't put an undue burden on Coq developers).
```
This sentence is very difficult to understand as given. Here is my interpretation and suggestion.

File: dev/ci/README-users.md
Comment (SkySkimmer) [reply dev/ci/README-users.md:15]: That would be better done by adding to the bench not just CI.

We already say

>You may also be interested in having your project tested in our
performance benchmark. Currently this is done by providing a `.dev` OPAM package
in https://github.com/coq/opam-coq-archive and opening an issue at
https://github.com/coq/coq/issues.

**Comment by SkySkimmer** [comment]
changed

**Comment by andres-erbsen** [comment]
This looks good to me. Thanks!

File: dev/ci/README-users.md
Comment (ejgallego) [reply dev/ci/README-users.md]: @palmskog ideally CI and bench would be fully integrated indeed.

**Review by proux01 — APPROVED** [review]
This seems to match what was agreed during yesterday's Coq call, I'll merge by the end of the week if no further comment

**Comment by proux01** [comment]
@coqbot merge now

**Pull request merged**
